zebra-chain/src/types.rs

```diff
@@ -15,26 +17,67 @@ use crate::serialization::{
     ReadZcashExt, SerializationError, WriteZcashExt, ZcashDeserialize, ZcashSerialize,
 };
 
-/// A 4-byte checksum using truncated double-SHA256 (two rounds of SHA256).
-#[derive(Copy, Clone, Eq, PartialEq)]
-pub struct Sha256dChecksum(pub [u8; 4]);
+/// A runtime validated type for representing amounts of zatoshis
+#[derive(Debug, Eq, PartialEq, Clone, Copy, Serialize, Deserialize)]
+pub struct Amount<C = NegativeAllowed>(i64, PhantomData<C>);
 
-impl<'a> From<&'a [u8]> for Sha256dChecksum {
-    fn from(bytes: &'a [u8]) -> Self {
-        use sha2::{Digest, Sha256};
-        let hash1 = Sha256::digest(bytes);
-        let hash2 = Sha256::digest(&hash1);
-        let mut checksum = [0u8; 4];
-        checksum[0..4].copy_from_slice(&hash2[0..4]);
-        Self(checksum)
+#[cfg(test)]
+impl<C> Arbitrary for Amount<C>
+where
+    C: AmountConstraint + fmt::Debug,
+{
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        C::valid_range().prop_map(|v| Self(v, PhantomData)).boxed()
```

Comment:
Question: Should `arbitrary_with()` reject `Amount`s outside the valid range?

Should this code use `try_from()`, so all arbitrary amounts are valid?
If there is an issue here, why didn't our tests catch it?

Reply:
I don't think its necessary because this is generating the amounts _from_ the valid range, so its impossible for a subsequent `try_from` to fail by construction. We could use try_from to do the construction here but I don't see any proptest native way of dealing with fallible sources for `Strategy` impls, no `prop_try_map` or similar, so we would have to unwrap inside of the `arbitrary_with` impl. It should all optimize away but I don't think its necessary in the first place and strikes me as excessively defensive programming.

Reply:
Sounds good. I didn't know that input range was limited.